### PR TITLE
Market Discovery Protocol with cyclic broadcasting

### PR DESCRIPTION
**Pull request by nieznanysprawiciel**

### CODE REVIEW

File: core/market/decentralized/src/matcher/store.rs [hunk 1]
```diff
@@ -218,4 +253,37 @@ impl SubscriptionStore {
...
+                                e
+                            );
+                            None
+                        }
+                    }
+                }
+            })
+            .collect::<Vec<SubscriptionId>>()
+            .await)
+    }
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/matcher/store.rs:288]: Use batch_select. Make set difference.

File: core/market/decentralized/src/matcher.rs [hunk 2]
```diff
@@ -140,60 +181,334 @@ impl Matcher {
...
+    _caller: String,
+    msg: GetOffers,
+) -> Result<Vec<Offer>, DiscoveryRemoteError> {
+    match resolver.store.get_offers_batch(msg.offers).await {
+        Ok(offers) => Ok(offers),
+        Err(e) => {
+            log::error!("Failed to get batch offers. Error: {}", e);
+            // TODO: Propagate error.
+            Ok(vec![])
+        }
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/matcher.rs:288]: Return error instead of empty vector

File: core/market/decentralized/src/matcher.rs [hunk 3]
```diff
@@ -140,60 +181,334 @@ impl Matcher {
...
+            let num_ours_offers = our_offers.len();
+            let num_to_broadcast = matcher.config.discovery.num_broadcasted_offers;
+
+            let all_offers = matcher
+                .store
+                .get_offers(None)
+                .await?
+                .into_iter()
+                .map(|offer| offer.id)
+                .collect::<Vec<SubscriptionId>>();
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/matcher.rs:377]: Don't query full Offers from database if we only need ids

File: core/market/decentralized/src/config.rs [hunk 4]
```diff
@@ -0,0 +1,48 @@
+use chrono::Duration;
+
+/// TODO: Decide where should this config be loaded from.
+///  We could deserialize it from .json file or use structopt and
+///  configure market through env variables.
+pub struct Config {
```
Comment (tworec) [inline core/market/decentralized/src/config.rs:6]: @jiivan pls look at this. We (with @nieznanysprawiciel ) would like to introduce some kind of marriage of this approach and yours from #453

File: core/market/decentralized/src/matcher/store.rs (hunk 1 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/matcher/store.rs:288]: After rethinking: I can't use batch_select. It doesn't return unsubscribed and expired Offers. So if I remove from `offers` result of batch_select, it won't be what I expect here

File: core/market/decentralized/src/db/model/subscription_id.rs [hunk 5]
```diff
@@ -173,6 +174,25 @@ impl From<SubscriptionParseError> for ErrorMessage {
...
+/// Pretty display vector, as list with all elements in separate row.
+/// Implementation uses Display trait for all elements (not Debug) to make
+/// them more readable.
+pub struct DisplayVec<'a, T>(pub &'a Vec<T>);
+
+impl<'a, T> Display for DisplayVec<'a, T>
+where
+    T: Display,
+{
+    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
```
Comment (tworec) [inline core/market/decentralized/src/db/model/subscription_id.rs:186]: there is https://docs.rs/itertools/0.8.2/itertools/trait.Itertools.html#method.format_with which is already used by our deps. so adding this dep will not increase resulting binary size most probably

**Review by tworec — COMMENTED** [review]
partial reviw

File: core/market/decentralized/src/matcher.rs [hunk 6]
```diff
@@ -140,60 +181,334 @@ impl Matcher {
...
+        .choose_multiple(&mut rand::thread_rng(), num_to_select)
+        .cloned()
+        .collect::<Vec<SubscriptionId>>();
+    random_offers.extend(our_offers);
+    random_offers
+}
+
+fn randomize_interval(mean_interval: std::time::Duration) -> std::time::Duration {
+    let mut rng = rand::thread_rng();
+    (2 * mean_interval).mul_f64(rng.gen::<f64>())
```
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/matcher.rs:468]: But look. We want to have mean interval in config, because it is human intuitive.
And this is how you compute random time from mean...

File: core/market/decentralized/src/db/dao/offer.rs [hunk 7]
```diff
@@ -53,6 +53,23 @@ impl<'c> OfferDao<'c> {
         .await
     }
 
+    pub async fn batch_select(
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/db/dao/offer.rs:56]: Or even `Vec<node_id>` with similar semantic...

File: core/market/decentralized/src/identity.rs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    async fn list(&self) -> Result<Vec<NodeId>, IdentityError> {
+        Ok(bus::service(identity::BUS_ID)
+            .send(identity::List {})
+            .await
+            .map_err(|e| IdentityError::GsbError(e.to_string()))?
+            .map_err(|e| IdentityError::ListError(e.to_string()))?
+            .iter()
+            .map(|identity_info| identity_info.node_id)
```
Comment (tworec) [inline core/market/decentralized/src/identity.rs:49]: what about locked identities? We should probably filter them out. 

I'm not sure if default identity can be locked. Probably yes... In such a case we should cease broadcasting for it probably

File: core/market/decentralized/src/matcher/store.rs [hunk 9]
```diff
@@ -92,6 +111,18 @@ impl SubscriptionStore {
             .collect())
     }
 
+    pub async fn get_offers_batch(
+        &self,
+        ids: Vec<SubscriptionId>,
```
Comment (tworec) [inline core/market/decentralized/src/matcher/store.rs:116]: this should be also more generic if you'll decide to follow my comment on `OffersDao`

File: core/market/decentralized/src/market.rs [hunk 10]
```diff
@@ -114,7 +119,11 @@ impl MarketService {
     }
 
     pub async fn get_offers(&self, id: Option<Identity>) -> Result<Vec<Offer>, MarketError> {
-        Ok(self.matcher.store.get_offers(id).await?)
+        Ok(self
```
Comment (tworec) [inline core/market/decentralized/src/market.rs:122]: I'm more and more concerned about why do we need those wrappers here in main market module. They should be covered within rest_api module. If you'll agree than please create issue based on that comment, as this should not be part of this PR

File: core/market/decentralized/src/matcher.rs [hunk 11]
```diff
@@ -83,13 +103,21 @@ impl Matcher {
         offer: &ClientOffer,
         id: &Identity,
     ) -> Result<Offer, MatcherError> {
-        // TODO: Handle broadcast errors. Maybe we should retry if it failed.
         let offer = self.store.create_offer(id, offer).await?;
         self.resolver.receive(&offer);
 
+        log::info!(
+            "Subscribed new Offer: [{}] using identity: {} [{}]",
```
Comment (tworec) [inline core/market/decentralized/src/matcher.rs:110]: 👍 good to have this in logs

File: core/market/decentralized/src/matcher/store.rs [hunk 12]
```diff
@@ -218,4 +253,37 @@ impl SubscriptionStore {
             false => Err(DemandError::NotFound(demand_id.clone())),
         }
     }
+
+    /// Returns Offers SubscriptionId from vector, that don't exist in our database.
+    pub async fn filter_existing(
```
Comment (tworec) [inline core/market/decentralized/src/matcher/store.rs:258]:
```suggestion
    pub async fn filter_out_existing(
```

File: core/market/decentralized/src/identity.rs (hunk 8 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/identity.rs:49]: Are you sure we want to cease broadcasting for locked identities?
That means that Offers from locked identities shouldn't appear on other Nodes? I think it's impossible.

Probably we want to disable negotiations for locked identities (and what more important signing Agreements). But still I don't think, we should filter these identities here

File: core/market/decentralized/src/protocol/discovery.rs [hunk 13]
```diff
@@ -52,60 +58,81 @@ pub struct Discovery {
...
+        // TODO: We shouldn't use send_as. Put identity inside broadcasted message instead.
+        let _ = bus::service(broadcast::BUS_ID)
+            .send_as(default_id, bcast_msg) // TODO: should we send as our (default) identity?
             .await?;
         Ok(())
     }
 
-    pub async fn broadcast_unsubscribe(
+    /// Ask remote Node for specified Offers.
+    pub async fn get_offers(
```
Comment (tworec) [inline core/market/decentralized/src/protocol/discovery.rs:92]: to indicate it is remote operation, not local store select
```suggestion
    pub async fn get_offers_from(
```

File: core/market/decentralized/src/matcher.rs [hunk 14]
```diff
@@ -140,60 +181,334 @@ impl Matcher {
...
+        Ok(our_offers)
     }
 }
 
-pub(crate) async fn on_offer_received(
+// =========================================== //
+// Discovery protocol messages handlers
+// =========================================== //
+
+pub(crate) async fn on_offer_ids_received(
```
Comment (tworec) [inline core/market/decentralized/src/matcher.rs:226]: misleading name.  Below is my proposition. more descriptive one
```suggestion
pub(crate) async fn filter_known_offer_ids(
```

File: core/market/decentralized/src/matcher.rs [hunk 15]
```diff
@@ -140,60 +181,334 @@ impl Matcher {
...
-            SaveOfferError::Expired(_) => Ok(Propagate::No(Reason::Expired)),
-            // Below errors are not possible to get from checked_store_offer
-            SaveOfferError::SaveError(_, _)
-            | SaveOfferError::SubscriptionValidation(_)
-            | SaveOfferError::WrongState { .. } => {
-                Ok(Propagate::No(Reason::Error(format!("{}", e))))
+        .map_err(|e| log::warn!("Error filtering Offers. Error: {}", e))?)
+}
+
+pub(crate) async fn on_offers_received(
```
Comment (tworec) [inline core/market/decentralized/src/matcher.rs:241]: same here - maybe?
```suggestion
pub(crate) async fn receive_offers_to_propagate(
```

**Review by tworec — COMMENTED** [review]
more comments

File: core/market/decentralized/src/db/dao/offer.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/db/dao/offer.rs:56]: I have to do 2 methods, because `get_offer` returns `OfferState` not `Offer`, so I can't write batch version that returns `Vec<OfferState>`.

File: core/market/decentralized/src/protocol/discovery.rs (hunk 13 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/protocol/discovery.rs:92]: Hmmm I think it's obvious. You call these method on Discovery interface and pass NodeId in parameter. I don't think it can be missinterpreted

File: core/market/decentralized/src/matcher.rs (hunk 15 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/matcher.rs:241]: For me your proposal is more misleading. Propagation isn't main goal of this callback, but name suggests it.

File: core/market/decentralized/src/market.rs (hunk 10 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/market.rs:122]: I think, having it on market level isn't something bad. It is better api if you call `get_offers` instead dispatching `matcher.store...`.
But if you change it in future, I don't mind.

File: core/market/decentralized/src/matcher.rs (hunk 3 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/matcher.rs:377]: Added comment for future

File: core/market/decentralized/src/db/model/subscription_id.rs (hunk 5 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/db/model/subscription_id.rs:186]: I used this formatter within `DisplayVec` implementation. If you meant to use it not here, but in places I display this vector, I'm not sure it would be a good idea, because it seems to be to long

File: core/market/decentralized/src/matcher/store.rs (hunk 9 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/matcher/store.rs:116]: I think the opposite. The bottom layer (Dao) should have the most generic functions as we can make. Top layers like Store should have specialized functions that hide unused details.
It makes `Store` a little bit more useful

File: core/market/decentralized/src/db/model/subscription_id.rs (hunk 5 above)
Comment (tworec) [reply core/market/decentralized/src/db/model/subscription_id.rs:187]: That's what I ment. Thx

File: core/market/decentralized/src/identity.rs (hunk 8 above)
Comment (tworec) [reply core/market/decentralized/src/identity.rs:49]: I've performed a test using locked identity via yapapi and it has passed. So locked identities are... not locked in sense I'd expect. Let's leave it for now.

I'd expect that locked identity does not allow usage REST APIs at all, ie. no subscriptions, no activities, no debit notes nor invoices. But thats sth which must be discussed on higher level than just Marketplace.

Comment (tworec) [reply core/market/decentralized/src/identity.rs:49]: Referenced this comment in #529

File: core/market/decentralized/src/matcher/store.rs (hunk 9 above)
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:116]: 👍 agreed

File: core/market/decentralized/src/identity.rs (hunk 8 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/identity.rs:49]: I agree to everything you said, that shouldn't be possible ;)
But I don't think, we shouldn't broadcast Offers, if they were already there.
To discuss later.

File: core/market/decentralized/src/market.rs (hunk 10 above)
Comment (tworec) [reply core/market/decentralized/src/market.rs:122]: ok, I've made an issue for it #530

File: core/market/decentralized/src/protocol/discovery.rs (hunk 13 above)
Comment (tworec) [reply core/market/decentralized/src/protocol/discovery.rs:95]: It was not to me. 
I propose below naming scheme for this file:
* sender `broadcast_offers` and its receiver `on_broadcast_offers`
* `broadcast_unsubscribes` and `on_broadcast_unsubscribes`
* `retrieve_offers` --> `on_retrieve_offers`

`get_offers` resembles local db operation rather than remote

Comment (tworec) [reply core/market/decentralized/src/protocol/discovery.rs:95]: And consequently `OffersReceived` rename to `OffersRetrieved` as it reflects this operation is triggered by us and not via broadcast.

Comment (nieznanysprawiciel) [reply core/market/decentralized/src/protocol/discovery.rs]: Fixed

File: core/market/decentralized/src/matcher.rs (hunk 15 above)
Comment (tworec) [reply core/market/decentralized/src/matcher.rs:242]: Ok, my proposition is far from good ;) 

But my point is naming those fns as `on_*` suggests that they are handlers, and they are, but not directly. Their handler nature is depicted via inputting them to the `add_data_handler`.

To be consistent with above I'd rename:
*  `on_offers_received` to `save_and_match_offers` and  
* `filter_known_offer_ids` to `filter_out_known_offer_ids`.
* `on_get_offers` to simply `get_offers`
* `on_offer_unsubscribed` to `unsubscribe_offers`

This will give an idea of what are those fns purposes.
Current names only depicts **how** they are used, and not **what for**.

Comment (tworec) [reply core/market/decentralized/src/matcher.rs:242]: I'm preparing PR based on this PR with my proposals. They are to broad to be contained within comments

Comment (tworec) [reply core/market/decentralized/src/matcher.rs:242]: see #531

File: core/market/decentralized/src/matcher/store.rs (hunk 1 above)
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs]: it was a pain for me that we do it that not optimal, so I've rewritten to use single select instead of multiple selects.
It was a broader effort of rewriting OfferDao

File: core/market/decentralized/src/matcher.rs (hunk 3 above)
Comment (tworec) [reply core/market/decentralized/src/matcher.rs]: fixed

File: core/market/decentralized/src/matcher.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher.rs]: Every positive number here will result with proper mean value when you are using standard distribution.

File: core/market/decentralized/src/protocol/discovery.rs (hunk 13 above)
Comment (tworec) [reply core/market/decentralized/src/protocol/discovery.rs]: rethinked again and also fixed

File: core/market/decentralized/src/matcher.rs (hunk 15 above)
Comment (tworec) [reply core/market/decentralized/src/matcher.rs]: fixed

**Review by tworec — APPROVED** [review]
